app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt

```diff
@@ -0,0 +1,280 @@
+            sendCrawlRequest(peer, block.publicKey.toByteArray(), Math.max(TrustChainBlock.GENESIS_SEQ, block.sequenceNumber - 5))
+        } else {
+            val signBlock = signBlock(peer, block)
+            val returnTrustChainBlock = ChainService.PeerTrustChainBlock.newBuilder().setPeer(me).setBlock(signBlock).build()
+            responseObserver.onNext(returnTrustChainBlock)
+            responseObserver.onCompleted()
+        }
+    }
+
+
+    override fun getPublicKey(request: ChainService.Empty, responseObserver: StreamObserver<ChainService.Key>) {
+        responseObserver.onNext(createPrivateKey())
+        responseObserver.onCompleted()
+    }
+
+    private fun createPrivateKey() =
+            ChainService.Key.newBuilder().setPublicKey(ByteString.copyFrom(keyPair.public.encoded)).build()
+
+    override fun sendLatestBlocks(request: ChainService.CrawlResponse, responseObserver: StreamObserver<ChainService.Key>) {
+        for (trustChainBlock in request.blockList) {
+            saveBlock(request.peer, trustChainBlock)
+        }
+
+        responseObserver.onNext(createPrivateKey())
+        responseObserver.onCompleted()
+    }
+
+    fun saveBlock(request: ChainService.PeerTrustChainBlock): ValidationResult? = saveBlock(request.peer, request.block)
+
+    fun saveBlock(peer: ChainService.Peer, block: MessageProto.TrustChainBlock): ValidationResult? {
+        Log.i(TAG, "Received half block from peer with IP: " + peer.hostname + ":" + peer.port +
+                " and public key: " + Peer.bytesToHex(peer.publicKey.toByteArray()))
+
+        val validation: ValidationResult
+        try {
+            validation = TrustChainBlock.validate(block, dbHelper)
+        } catch (e: Exception) {
+            e.printStackTrace()
+            return null
+        }
+
+        Log.i(TAG, "Received block validation result " + validation.toString() + "("
+                + TrustChainBlock.toString(block) + ")")
+
+        if (validation.getStatus() === ValidationResult.ValidationStatus.INVALID) {
+            for (error in validation.getErrors()) {
+                Log.e(TAG, "Validation error: " + error)
+            }
+            return null
+        } else {
+            dbHelper.insertInDB(block)
+        }
+
+        val pk = keyPair.public.encoded
+        // check if addressed to me and if we did not sign it already, if so: do nothing.
+        if (block.linkSequenceNumber != TrustChainBlock.UNKNOWN_SEQ ||
+                !Arrays.equals(block.linkPublicKey.toByteArray(), pk) ||
+                null != dbHelper.getBlock(block.linkPublicKey.toByteArray(),
+                        block.linkSequenceNumber)) {
+            Log.e(TAG, "Received block not addressed to me or already signed by me.")
+            return null
+        }
+
+        // determine if we should sign the block, if not: do nothing
+        if (!shouldSign(block)) {
+            Log.e(TAG, "Will not sign received block.")
+            return null
```

Comment:
Would it be a good idea to give a validationstatus == `wontsign` or something?

Reply:
I aim to refactor that later when we actually start adding our own business logic.